proposals/0031-hlsl-vector-matrix-operations.md

```diff
@@ -156,12 +144,500 @@ vector<float, 3> ApplyNeuralMaterial(vector<half, 8> inputVector) {
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename BufferTy, DataType DT, uint M, uint K, MatrixLayout ML,
+          bool Transpose>
+struct MatrixRefImpl {
+  BufferTy Buffer;
+  uint StartOffset;
+  uint Stride;
+};
+
+template <DataType DT, uint M, uint K, MatrixLayout ML, bool Transpose = false>
+using MatrixRef = MatrixRefImpl<ByteAddressBuffer, DT, M, K, ML, Transpose>;
+
+template <DataType DT, uint M, uint K, MatrixLayout ML, bool Transpose = false>
+using RWMatrixRef = MatrixRefImpl<RWByteAddressBuffer, DT, M, K, ML, Transpose>;
+
+} // namespace linalg
+} // namespace dx
+```
+
+### struct VectorRef
+
+`VectorRef` and `RWVectorRef` specify a reference to a vector in memory along
+with the format of each element.
+
+As with `MatrixRef`, two versions are provided - one for vectors stored in
+`ByteAddressBuffer` and another for `RWByteAddressBuffer`. A base class,
+`VectorRefImpl`, covers both of them.
+
+Example usage:
+
+```c++
+ByteAddressBuffer ROBuffer;
+RWByteAddressBuffer RWBuffer;
+
+void Example() {
+  using namespace dx::linalg;
+
+  VectorRef<DATA_TYPE_FLOAT16> VectorA = {ROBuffer, /*offset=*/128};
+  VectorRef<DATA_TYPE_FLOAT32> VectorB = {ROBuffer, /*offset=*/128};
+  RWVectorRef<DATA_TYPE_SINT16> VectorC = {RWBuffer, /*offset=*/64};
+}
+```
+
+Template parameter:
+
+- `DT` - the data type of each element stored in the buffer
+
+Members:
+
+- `Buffer` - the buffer that the vector is stored in
+- `StartOffset` - the offset, in bytes, from the beginning of the buffer to
+  where the vector is located.
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename BufferTy, DataType DT> struct VectorRefImpl {
+  BufferTy Buffer;
+  uint StartOffset;
+};
+
+template <DataType DT> using VectorRef = VectorRefImpl<ByteAddressBuffer, DT>;
+
+template <DataType DT>
+using RWVectorRef = VectorRefImpl<RWByteAddressBuffer, DT>;
+
+} // namespace linalg
+} // namespace dx
+
+```
+
+### struct Vector
+
+> NOTE: it's possible that one resolution of [441] may remove the need for this
+> type entirely.
+
+The `Vector` struct is a wrapper around `vector`, adding an interpretation value
+that controls how the data in the vector should be interpreted. Although the
+struct can be used directly, it is likely more ergonomic to use the
+`InterpretedVector` function that's also described here.
+
+Example usage:
+
+```c++
+ByteAddressBuffer Buffer;
+void Example() {
+  using namespace dx::linalg;
+
+  MatrixRef<DATA_TYPE_FLOAT16, 128, 128, MATRIX_LAYOUT_MUL_OPTIMAL, true>
+      Matrix = {Buffer, 0, 0};
+
+  vector<float, 128> V = 0;
+  vector<float, 128> Result =
+      Mul<float>(Matrix, InterpretedVector<DATA_TYPE_E4M3>(V));
+
+  // alternative:
+  Vector<float, 128, DATA_TYPE_E4M3> IV = {V};
+  vector<float, 128> Result2 = Mul<float>(Matrix, IV);
+}
+```
+
+Implementation:
+
+```c++
+namespace dx {
+namespace linalg {
+
+template <typename T, int N, DataType DT> struct Vector {
```

Comment:
Is DT used?

Reply:
Yes, but not here.  It's matched by a template parameter to `Mul` or `MulAdd`.